microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/jaxrs/java/JaxRsWorkspaceSymbolParticipant.java

```diff
@@ -57,196 +45,90 @@ public class JaxRsWorkspaceSymbolParticipant implements IJavaWorkspaceSymbolsPar
 			}
-			return new Location(uri, range);
-		} catch (JavaModelException e) {
-			return null;
-		}
+		});
 	}
 
-	private static String getEndpointUriFromAnnotatable(IAnnotatable annotatable, String applicationPrefix) {
-		StringBuilder builder = new StringBuilder();
-		IAnnotatable current = annotatable;
-
-		while (current != null && current != (IAnnotatable) ((IJavaElement) current).getAncestor(IJavaElement.TYPE)) {
-			if (hasRestClientAnnotation(current)) {
-				// This is not an endpoint, it's an interface to an outside service
-				return null;
+	/**
+	 * Returns the provider that can provide JAX-RS method info for the given class,
+	 * or null if no provider can provide info.
+	 *
+	 * @param typeRoot the class to collect JAX-RS method info for
+	 * @param monitor the progress monitor
+	 * @return the provider that can provide JAX-RS method info for the given class,
+	 *         or null if no provider can provide info
+	 */
+	private static IJaxRsInfoProvider getProviderForType(ITypeRoot typeRoot, IProgressMonitor monitor) {
+		for (IJaxRsInfoProvider provider : JavaFeaturesRegistry.getInstance().getJaxRsInfoDefinitions()) {
+			if (provider.canProvideJaxRsMethodInfoForClass(typeRoot, monitor)) {
+				return provider;
 			}
-			prependPathSegment(builder, current);
-			current = (IAnnotatable) ((IJavaElement) current).getAncestor(IJavaElement.TYPE);
 		}
-		if (hasRestClientAnnotation(current)) {
-			return null;
+		IJaxRsInfoProvider defaultProvider = DefaultJaxRsInfoProvider.getInstance();
+		if (defaultProvider.canProvideJaxRsMethodInfoForClass(typeRoot, monitor)) {
+			return defaultProvider;
 		}
-		if (current != null) {
-			prependPathSegment(builder, current);
-		}
-		if (applicationPrefix != null) {
-			builder.insert(0, applicationPrefix);
-			if (applicationPrefix.charAt(0) != '/') {
-				builder.insert(0, '/');
-			}
-		}
-		return builder.toString();
+		LOGGER.severe("Attempted to collect JAX-RS info for " + typeRoot.getElementName()
+				+ ", but no participant was suitable, despite the fact that an earlier check found a suitable participant");
+		return null;
 	}
 
-	private static void prependPathSegment(StringBuilder builder, IAnnotatable current) {
-		String pathValue = null;
-		try {
-			pathValue = JaxRsUtils.getJaxRsPathValue(current);
-		} catch (JavaModelException e) {
-		}
-		if (pathValue != null) {
-			builder.insert(0, pathValue);
-			if (pathValue.charAt(0) != '/') {
-				builder.insert(0, "/");
+	private static Set<ITypeRoot> getAllJaxRsTypes(IJavaProject javaProject, IProgressMonitor monitor) {
+		Set<ITypeRoot> jaxrsTypes = new HashSet<>();
+		for (IJaxRsInfoProvider provider : JavaFeaturesRegistry.getInstance().getJaxRsInfoDefinitions()) {
+			jaxrsTypes.addAll(provider.getAllJaxRsClasses(javaProject, monitor));
+			if (monitor.isCanceled()) {
+				return null;
 			}
 		}
+		jaxrsTypes.addAll(DefaultJaxRsInfoProvider.getInstance().getAllJaxRsClasses(javaProject, monitor));
```

Comment:
To avoid having specific code with DefaultJaxRsInfoProvider here, I will add new DefaultJaxRsInfoProvider() at the end of the list of getJaxRsInfoDefinitions once all IJaxRsInfoProviders are loaded by extension point.